### ISSUE TITLE
PShape.getVertexCount() always outputs 0

### ISSUE THREAD

**Issue by dtplsongithub** [processing/processing4#896]
### Most appropriate sub-area of Processing 4?

Core/Environment/Rendering

### Processing version

4.3.1

### Operating system

Windows 11 (24H2)

### Steps to reproduce this

`PShape.getVertexCount()` always seems to output 0. (tried 1 .svg file and 4 .obj files, even included the .mtl files)

### snippet

```processing
PShape shape;
  
void setup() {
  size(640, 360, P3D);
    
  shape = loadShape("cuboctahedron.obj");
  println(shape.getVertexCount()); // zero???
}

void draw() {
  background(0);
  lights();
  translate(width/2, height/2, -200);
  scale(50);
  rotateY((float)millis()/2000);
  shape(shape);
}
```

**Comment by hx2A** [comment]
@dtplsongithub , can you check if your shape is a group shape?

I just tested this on the Processing example "LoadDisplayOBJ", which shows a rotating rocket. The vertex count was zero but I could use the below code to get the vertices for the object's children:

```java
  rocket = loadShape("rocket.obj");
  println(rocket.getVertexCount());
  int count = 0;
  for (PShape shape : rocket.getChildren()) {
    count += shape.getVertexCount();
  }
  println(count);
```

The output is:

```txt
0
1101
```

I would argue that a call to `getVertexCount()` on a group shape should return the total vertex count of its children, and that all renderers (JAVA2D, P2D, P3D) should behave in a similar fashion for loaded shapes.

**Comment by dtplsongithub** [comment]
the code you provided me works, i guess i was using the method wrong...

**Comment by hx2A** [comment]
> the code you provided me works, i guess i was using the method wrong...

No, I think the code you originally wrote should work. `getVertexCount()` is confusing if it returns zero for group shapes. Also, consider that the 1101 vertex count of my example could also be wrong. What if the group shape contains a mixture of mesh shapes and other group shapes? Group shapes can be nested. Some of the calls to `getVertexCount()` for the shape's children could also be zero. The 1101 value could be an under-count of the correct number.

Let's keep this open so we can use this as an opportunity to improve Processing. It would likely be an easy PR, so I'll add some tags so a first-time contributor can fix it.

**Comment by hx2A** [comment]
And @dtplsongithub , if you, or anyone else reading this would like some help working through the PR, I'm available to assist!

**Comment by Stefterv** [comment]
Adding a warning message, that groups are ignored in the count would also be helpful. Maybe even a `countGroups = false` parameter.

**Comment by micycle1** [comment]
@hx2A 
I think the PShape data structure is due an overhaul (perhaps one for Processing 5😄).
Beyond the confusion you raise, one even bigger problem is how holes are encoded. Vertices of holes are not directly accessible and must be computed from the PShape's vertex codes using some fairly involved logic.

For 2D shapes the best approach would be follow the model described by _[Well-known text representation](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry) for geometries_.

**Comment by hx2A** [comment]
> @hx2A I think the PShape data structure is due an overhaul (perhaps one for Processing 5😄). Beyond the confusion you raise, one even bigger problem is how holes are encoded. Vertices of holes are not directly accessible and must be computed from the PShape's vertex codes using some fairly involved logic.
> 
> For 2D shapes the best approach would be follow the model described by _[Well-known text representation](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry) for geometries_.

Well, much of PShape isn't that accessible. It's very much a custom data structure that has evolved over time to work well with Processing.

Overhauling PShape would probably need to wait for #881 as PShapeOpenGL is fairly complex and nobody should touch it if it is going to go away eventually.

I like the idea of using WKT for 2D geometry. There is a Java library called [JTS](https://github.com/locationtech/jts) that Processing could incorporate or at least learn from, and I believe it uses WKT. This library is closely related to [GEOS](https://github.com/libgeos/geos), which is used by the Python library [shapely](https://github.com/shapely/shapely).

**Comment by micycle1** [comment]
> I like the idea of using WKT for 2D geometry. There is a Java library called [JTS](https://github.com/locationtech/jts) that Processing could incorporate or at least learn from, and I believe it uses WKT. This library is closely related to [GEOS](https://github.com/libgeos/geos), which is used by the Python library [shapely](https://github.com/shapely/shapely).

It was within JTS that I came across that format!
I use JTS extensively in my geometry library for Processing and have routines for converting between PShapes and JTS geometries (if you're curious see [here](https://github.com/micycle1/PGS/blob/28fe7a9ff211a767c926465d84c56f57e4c48a85/src/main/java/micycle/pgs/PGS_Conversion.java#L354)). It was also through this exercise that I came across the poor design of PShape as a geometry data structure (notable issues are holes and affine transformations -- the effects of `scale()` etc that [still](https://github.com/processing/processing4/issues/413) aren't publicly accessible!).

**Comment by hx2A** [comment]
Hmmm, if PShape was more closely aligned with JTS, there are a lot of nice computational geometry features we could add to Processing.

**Comment by inteqam** [comment]
Hi @hx2A , I want to work on this bug , I’ve taken a look at the issue, but I’m not entirely sure how to approach solving it. Could you provide me with some key points or guidance on how to tackle this problem? Any insights or suggestions on where to start would be really helpful!

**Comment by inteqam** [comment]
Hi @hx2A , I want to fix this bug ,how should i approach this problem ? Any insights or suggestions on where to start would be really helpful!

**Comment by SableRaf** [comment]
Hi @inteqam and thanks for your interest in contributing to Processing 💙

Here are a few things you can do to get started:

1. [**Download Processing**](https://processing.org/download) and try to reproduce this issue.
2. **Set up the development environment** by following the [build instructions](https://github.com/processing/processing4/blob/main/build/README.md).
3. **Find where `PShape.getVertexCount()` is implemented** and consider the problem and discussion above.

@hx2A suggested:

> I would argue that a call to getVertexCount() on a group shape should return the total vertex count of its children, and that all renderers (JAVA2D, P2D, P3D) should behave in a similar fashion for loaded shapes.

How would you go about this?

Once you've done this, feel free to ask if you have specific questions. We're happy to help!

**Comment by hx2A** [comment]
Hi @inteqam ! Thank you for your interest in pursing this bug.

I've been swamped and can't reply to things as quickly.

The basic problem with `getVertexCount()` is that it returns 0 for `GROUP` shapes. What it can do instead is first check to see what shape type it is. If it is not a `GROUP`, it can get the count as it does now. If it is a `GROUP` shape, it should iterate through the children and return the sum of the childrens' `getVertexCount()` values. Of course some of those children can also be `GROUP` shapes. If that's the case, recursion will happen and the user should still get the correct result.

You might need to test and implement this fix in more than one place. The OpenGL renderers `P2D` and `P3D` have a `PShapeOpenGL` class.  There's also a `PShapeJava2D` class, `PShapeSVG` class, and `PShapeOBJ` class, in addition to the `PShape` class they all ultimately inherit from. Trace through the code for each these classes to look for `getVertexCount()` methods that need to be changed.

I suggest that after you get the development environment working, as @SableRaf suggested, you build a few test Sketches to verify and explore the problem. Then you can start fiddling with the code to understand how it works and explore solutions.

**Comment by inteqam** [comment]
Thanks @hx2A and @SableRaf  for taking the time to explain!                                                                                                                            
                                                                                                                                                                                                               
 After reproducing this issue I have traced the code for each classes and I found `getVertexCount()` method in `PShape`  and `PShapeOpenGL` only . I have updated `getVertexCount()` method to count Vertex of Group Shapes using recursion ensuring that no Group is missed in Nested Group case       
Updated `getVertexCount()`  :                                                                                                                                                                      
 `PShape` : 

![Image](https://github.com/user-attachments/assets/6eba086f-c4eb-476e-a52b-428bba89dea9)                                                 
 and    `PShapeOpenGL` :  

![Image](https://github.com/user-attachments/assets/c00fac37-80e0-4ce5-9d63-71b564a2d885)

After Updating  I have tested `getVertexCount()` for `svg` and `obj` files , which are now succesfully counting the vertex count of its children : 

![Image](https://github.com/user-attachments/assets/becbd868-7f2b-44db-a8a4-1cf5ebd37cfe)

Is there anything important I should consider or might have missed? For now, I'm adding it to the PR.

**Comment by Junology** [comment]
Sorry for interruption, but I am afraid that some people may expect `getVertexCount()` to return the valid upper bound for the indices passed to vertex query functions such as `getVertex()`.
Namely, the change may cause `ArrayIndexOutOfBoundsException` if one accidentally pass a `GROUP` shape to a vertex modifying function like
```processing
void disturbShape(PShape shape, float amp) {
  int vcount = shape.getVertexCount();
  PVector vec = new PVector();
  for(int i = 0; i < vcount; ++i) {
    shape.getVertex(i, vec);
    vec.x += randomGaussian() * amp;
    vec.y += randomGaussian() * amp;
    vec.z += randomGaussian() * amp;
    shape.setVertex(i, vec);
  }
}
```
Also, be aware that the meaning of *"vertices"* changes in the "Tessellation Update Mode"; see my comments in #902.

**Comment by Junology** [comment]
I would rather suggest implementing the feature of counting all the vertices as a new function, e.g. as `getVertexTotalCount()`, instead of refactoring existing functions.
In that case, it will be enough to write the function only in `PShape` class; if it is written using `getVertexCount()`, it would work correctly in any derived classes.

**Comment by hx2A** [comment]
> but I am afraid that some people may expect `getVertexCount()` to return the valid upper bound for the indices passed to vertex query functions such as `getVertex()`.

@Junology , this is an excellent point. A common use case of `getVertexCount()` is for loops. This PR would change the method so it can't be used that way.

> I would rather suggest implementing the feature of counting all the vertices as a new function, e.g. as `getVertexTotalCount()`, instead of refactoring existing functions.
> In that case, it will be enough to write the function only in `PShape` class; if it is written using `getVertexCount()`, it would work correctly in any derived classes.

I do like this suggestion. We can add a new method that recurses through the PShape's children to get the total number of vertices in the shape. @SableRaf and @Stefterv , what is the process for proposing new methods? Do you think it is worth it in this case?

**Comment by Stefterv** [comment]
I agree that it should not be changing existing functionality, especially regarding the use case @Junology provided. I wonder if we can instead add an extra optional parameter to the function, e.g. `getVertex(boolean includeChildren = false)` that will make sure the reference of `getVertexCount` shows the option to also count the children

**Comment by SableRaf** [comment]
Thanks everyone for your thoughtfulness 😃 Overall, I think the solution proposed by @stefterv makes the most sense for its simplicity, and discoverability of the feature.

Bonus points for requiring only minor changes to @inteqam's existing PR 😉

@hx2A Thanks for your question. I don’t think we have a strict process for adding functions, methods, or classes to the API. It’s mostly based on discussion and community consensus. Backwards compatibility is a big priority, so anything that breaks or changes existing functionality would need more justification. We also try to keep the core small and maintainable, so small, local changes are easier to approve than larger ones, or even small ones that have a broader impact on functionality.

**Comment by hx2A** [comment]
> I wonder if we can instead add an extra optional parameter to the function, e.g. `getVertex(boolean includeChildren = false)` that will make sure the reference of `getVertexCount` shows the option to also count the children

Excellent idea, @Stefterv ! That's perfect!

> It’s mostly based on discussion and community consensus.

I do like these discussions and taking the time to think through changes. It's so important for the long term success of the library.
